docs/rationale.md

```diff
@@ -0,0 +1,144 @@
+# Test262 maintenance rationale
+
+Explanations behind the practices promoted by the project maintainers.
+
+## Vestigial tests
+
+Test262 has been maintained for many years, and the practices used to write
+tests have evolved alongside the needs of its consumers. When conventions
+change, old tests are typically updated to accommodate the new practices. That
+doesn't always happen, though, and one doesn't have to look very far to find
+examples of tests which contradict the preferred patterns.
+
+For instance:
+
+- tests which expression expectations with `throw` statements inside of
+  conditional statements rather than the assertion API implemented by the
+  harness files (though this explicitness will always been desirable when
+  asserting the semantics of conditional statements and `throw` statements
+  themselves)
+- tests with file names derived from section numbers in the 5th edition of
+  ECMA262, e.g. `built-ins/Array/15.4.5-1.js`
+- tests which validate multiple behaviors, using elaborate comment blocks to
+  designate sections
+- tests which use deprecated harness functions, e.g. `verifyEnumerable`,
+  `verifyConfigurable`, and `verifyWritable`
+
+Since existing tests do not necessarily reflect the project's current
+best-practices, it's especially important for test authors to familiarize
+themselves with the contribution guidelines.
+
+## Test generation
+
+The project includes a software tool for generating test material from abstract
+templates. The generator was designed to promote uniformity of coverage,
+particularly for parts of the grammar that are used in many productions
+(specifically: the destructuring assignment patterns introduced in the 6th
+edition of ECMA262).
+
+This tool makes it easy to introduce enormous numbers of tests. Introducing
+more tests is not a goal unto itself, though. Test262 prioritizes coherence in
+its coverage of the specification, and it recognizes that the value of tests,
+as measured by their likelihood to identify defects, varies.
+
+For these reasons, the maintainers urge restraint in the application of the
+test generation tool.
+
+## File structure
+
+For practical reasons, tests are organized in a tree structure according to the
+conventions of modern file systems. Unfortunately, this structure is not
+expressive enough to model the semantics of a rich and evolving programming
+language like ECMAScript. This means the common and crucial task of coverage
+assessment will likely always be a challenge, but strong conventions around
+file organization can help.
+
+Tests for syntax-derived operations are organized according to the language
+grammar, with directories used to describe non-terminals. Tests for built-in
+APIs are organized according to the identifiers by which they can be accessed,
+with directories used to describe the sequence of properties that can be used
+from the global scope.
```

Comment:
I'd love to have this expanded so that it fully answers the question "Where do I put my test?" as in https://github.com/tc39/test262/pull/3406#discussion_r829390527
- For "the language grammar", including an example of the `language/statements/` directory with a link to https://tc39.es/ecma262/#sec-ecmascript-language-statements-and-declarations would help illustrate the correspondence that Rick showed in that other thread.
- The part about built-in identifiers could use an example for clarity
- We should say something about `intl402/` vs `built-ins/`, I'd guess it wouldn't be obvious to some newcomers that some hosts don't have Intl and toLocaleString
- For completeness we could mention `annexB/` and `harness/`

Reply:
Great ideas, all.